### DISCUSSION TITLE
Firing an event after a session is committed

### DISCUSSION THREAD

**Discussion by ay-ramp** [sqlalchemy/sqlalchemy#9300] [Usage Questions] [answered]
Hi there! Was wondering if there is a recommended approach for implementing something like django's `on_commit` in SqlAlchemy 1.4.

Ideal behavior would be something like:

```python

with session.begin():
  print("A")
  on_commit(lambda: print("OC-1"))
  print("B")
  with session.begin_nested():
    on_commit(lambda: print("OC-2"))
  print("C")
  try:
    with session.begin_nested():
      on_commit(lambda: print("not printed"))
      raise Exception()
  except: pass
```

yield:

```
A
B
C
OC-1
OC-2
```

It seems like the [after_commit](https://docs.sqlalchemy.org/en/14/orm/events.html#sqlalchemy.orm.SessionEvents.after_commit) hook is almost what i'm looking for, but haven't been able to get it to deal with nested transactions correctly (not firing on rollbacks + deferring execution until an actual `COMMIT` not just a savepoint release)

**Comment by zzzeek** [comment, marked as answer]
the [after_transaction_end](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.after_transaction_end) event might be more versatile in this regard, that will give you the session's view of both nested (i.e. savepoint) and non-nested (i.e. the outermost) transaction.

**Reply by ay-ramp** [reply]
thanks, that's helpful -- what would be the best way to determine if the transaction was rolledback or commited (savepoint or non) from that hook?

**Reply by zzzeek** [reply]
look at the .nested attribute for savepoints and .parent is None for the outermost real transaction.

**Reply by ay-ramp** [reply]
Ah sorry wasn't clear -- I do see how I can understand outermost / not here. What I'm not sure about is how to to tell from [after_transaction_end](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.after_transaction_end) whether the transaction ended due to a successful commit() or a rollback() -- if it's the former, i'd want to either continue staging functions enqueued/run them if it's the outermost; if it's the latter, i'd want to discard the enqueued functions

let me know if i'm overcomplicating this / you know of any prior art!

thank you 🙏

**Reply by zzzeek** [reply]
I dont think the transaction object itself tracks if it was rolled back or committed, so I would also use the after_commit / after_rollback hooks too!  the unfortunate part about these hooks is that they aren't directly passed the SessionTransaction object that's coordinating the commit/rollback, so it takes extra steps to link these events to the SessionTransaction passed to after_transaction_begin().   When the Session is handed to those event hooks, look at `Session.transaction`, that will give you the SessionTransaction object that's currently active.

API improvements that can be made here:

* Session.transaction is deprecated, in favor of Session.get_transaction() and Session.get_nested_transaction(), however, that means there's no easy way to get "the current SessionTransaction", so we should add another method Session.get_current_transaction().
* SessionTransaction doesn't store if it was committed, rolled back, or just closed.

**Reply by zzzeek** [reply]
I guess the total answer to your original question might be:

1. just use the after_commit() hook
2. inside the hook, check `not session.in_nested_transaction()`, that way you'll know this is not  a savepoint

**Reply by ay-ramp** [reply]
got it, sounds good. last question on this, is it kosher to do something like `current_transaction = session.get_nested_transaction() or session.get_transaction() `?

**Reply by zzzeek** [reply]
yes that would be fine, I was trying to save you on the verbosity of that :)
